pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -43,11 +46,48 @@ const (
 
 var runInTestMode bool
 var runInDemoMode bool
+var slowOperationBuckets = []float64{30, 90, 180, 360, 720, 1440}
+var reconcileCounters = prometheus.NewCounterVec(prometheus.CounterOpts{
+	Name: "metal3_reconcile",
+	Help: "The number of times hosts have been reconciled",
+}, []string{"host"})
```

Comment:
Shouldn't we also record the namespace for each of these metrics?

Reply:
We need a unique id for the host. We could include the namespace in the host string, but I don't think we want to use a separate label for it.

Reply:
If you want to record the namespace, I'd recommend to use a separate label. It's easier to work with in alerts/queries and from a cardinality standpoint, it will be the same as including the namespace in the host value.

Reply:
OK, Added a label for the namespace, but... reading the Prometheus docs for the first time [I see](https://prometheus.io/docs/practices/naming/#labels):

> Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values.

Does that mean that by including the Host name we're abusing Prometheus to try to do something that's really the job of logging? [Elsewhere](https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels) it says that 100k timeseries is manageable and we are unlikely to ever have that many Hosts, so maybe it's ok?

Reply:
The `host` label seems to be in the grey area between "fine" and "absolutely not fine". It can be ok for Prometheus to store a couple of metrics with a 10k cardinality but 1) it's probably better to avoid it if possible and 2) a Kubernetes cluster has so many components which makes it hard to know in advance how many metrics will have high-cardinality.

Usually I would look at how the metric is going to be used for alerting. With  `metal3_reconcile_total` and `metal3_reconcile_error_total`, you'll be able to alert on reconciliation errors: do you need the alert to be per-host, per-namespace or global?

Reply:
I wouldn't expect to have significantly more hosts than nodes in a cluster. There may be some unprovisioned hosts, but the point of adding them to a cluster is to provision them to become nodes so at some point the counts should be equal.

All of the hosts in an OpenShift cluster will be in the same namespace, although that won't automatically be true for other types of clusters.

I originally included host as a label because the metrics for nodes include the node id as a label. I expect alerts to be based on totals and then drilling in to the stats to expose which host(s) have failures. Is that not how the data is typically used for nodes?

Reply:
I can definitely see the value in having per-host metrics. If e.g. a single host is constantly reconciling then something may not be right with it, and you could monitor that and alert in a way that you probably couldn't if it was lost in all of the noise of every host reconciling.
What I think I'm [reading](https://prometheus.io/docs/practices/instrumentation/) about Prometheus though is that this isn't really the use case it's intended for; it's for monitoring the performance of the software itself (request processing time, query processing time, external call failures, unusual code paths taken, that kind of thing) rather than any specific user or set of data that it is operating on.
So maybe we should just go with global or per-namespace?

Reply:
Prometheus metrics are primarily for alerting. Without any label, you can still be notified that something is wrong and drill down to logs or events to determine whether it's a specific host which can't be reconciled or a more general issue.
Again it's ok to have a few metrics with large cardinality (hundred-ish to 1,000 values per label) but I'm worried that in the Kubernetes context, we already have too many of them...

Reply:
On this topic, the Kubernetes instrumentation SIG has put together some documentation:

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md

Reply:
I've read that documentation several times now and I still don't know what the answer is. My summary of it would be "we never do that, except for all the times it might have been tempting to do that, in which cases we totally did it".
The approach I went for at this stage is to keep the name+namespace labels for only a handful of metrics - the overall reconcile count, the on/off counts, and the histograms of the operation times. Nothing else uses those labels.